java/ql/automodel/src/AutomodelApplicationModeExtractNegativeExamples.ql

```diff
@@ -48,18 +48,20 @@ from
 where
   endpoint = getSampleForCharacteristic(characteristic, 100) and
   extensibleType = endpoint.getExtensibleType() and
+  // the node is know not to be an endpoint of any appropriate type
+  forall(EndpointType tp | tp = endpoint.getAPotentialType() |
+    characteristic.hasImplications(tp, false, _)
+  ) and
+  // the lowest confidence across all endpoint types should be at least highConfidence
+  confidence = min(float c | characteristic.hasImplications(endpoint.getAPotentialType(), false, c)) and
   confidence >= SharedCharacteristics::highConfidence() and
-  characteristic.hasImplications(any(NegativeSinkType negative), true, confidence) and
   meta.hasMetadata(endpoint, package, type, subtypes, name, signature, input, output, isVarargsArray) and
-  // It's valid for a node to satisfy the logic for both `isSink` and `isSanitizer`, but in that case it will be
-  // treated by the actual query as a sanitizer, since the final logic is something like
-  // `isSink(n) and not isSanitizer(n)`. We don't want to include such nodes as negative examples in the prompt, because
-  // they're ambiguous and might confuse the model, so we explicitly exclude all known sinks from the negative examples.
-  not exists(EndpointCharacteristic characteristic2, float confidence2, SinkType positiveType |
-    not positiveType instanceof NegativeSinkType and
+  // It's valid for a node to be both a potential source/sanitizer and a sink. We don't want to include such nodes
+  // as negative examples in the prompt, because they're ambiguous and might confuse the model, so we explicitly them here.
+  not exists(EndpointCharacteristic characteristic2, float confidence2, EndpointType type2 |
     characteristic2.appliesToEndpoint(endpoint) and
     confidence2 >= SharedCharacteristics::maximalConfidence() and
-    characteristic2.hasImplications(positiveType, true, confidence2)
+    characteristic2.hasImplications(type2, true, confidence2)
```

Comment:
Double-checking my own understanding: here, we need `characteristic2` to _not_ be equal to `characteristic` in the outer context.

This will be the case for a combination of two reasons:
 1. in L64, we require a `true` implication coming out of `characteristic2`, and above we've required a `false` implication coming out of `characteristic`; and
 2. we don't have any characteristics that are able to return a `true` some times and `false` other times.

If that's indeed a correct and complete assessment, I'd suggest still adding a check here like `characteristic != characteristic2 and` to guard against bullet 2. becoming invalid because someone might build an unusual characteristic in the future[^1]. We have to assume that the number of characteristics is open to extension.

[^1]: We had this logic before because we required `characteristic2` to be a positive type.